robotiq_hande_driver/hardware/include/robotiq_hande_driver/application.hpp

```diff
@@ -0,0 +1,315 @@
+     */
+    void initialize(double gripper_position_min, double gripper_position_max, std::string tty_port, int baudrate, char parity, int data_bits, int stop_bit, int slave_id) {
+        gripper_position_min_ = gripper_position_min;
+        gripper_position_max_ = gripper_position_max;
+        gripper_postion_step_ = (gripper_position_max_ - gripper_position_min_) / 255.0;
+        protocol_logic_.initialize(tty_port, baudrate, parity, data_bits, stop_bit, slave_id);
+    };
+
+    /**
+     * @brief Configures driver session.
+     * @return int, when error <0.
+     * @note The status should be checked to verify successful execution. An exception is thrown if communication issues occur.
+     */
+    int configure() {
+        int result;
+
+        result = protocol_logic_.configure();
+
+        return result;
+    };
+
+    /**
+     * @brief Deinitializes driver.
+     *
+     * @param none
+     * @return None.
+     * @note The status should be checked to verify successful execution. An exception is thrown if communication issues occur.
+     */
+    void cleanup() {
+        protocol_logic_.cleanup();
+    };
+
+    /**
+     * @brief Stops the gripper movement.
+     *
+     * @param none
+     * @return None.
+     * @note The status should be checked to verify successful execution. An exception is thrown if communication issues occur.
+     */
+    void stop() {
+        protocol_logic_.stop();
+    };
+
+    /**
+     * @brief Resets the gripper by deactivating and reactivating it.
+     *
+     * @param none
+     * @return None.
+     * @note The status should be checked to verify successful execution. An exception is thrown if communication issues occur.
+     */
+    void reset() {
+        protocol_logic_.reset();
+    };
+
+    /**
+     *  Emergency auto-release, gripper fingers are slowly opened, reactivation necessary
+     *
+     * @param none
+     * @return None.
+     * @note The status should be checked to verify successful execution. An exception is thrown if communication issues occur.
+     */
+    void auto_release() {
+        protocol_logic_.auto_release();
+    };
+
+    /**
+     * @brief Activates the gripper, making it ready for use.
+     *
+     * @param blocking If true wait until the gripper is active.
+     * @return None.
+     * @note The status should be checked to verify successful execution. An exception is thrown if communication issues occur.
+     */
+    void activate(bool blocking=true) {
+        int iter = 0;
+        read();
+
+        if (status_.is_ready)
+            printf("Gripper already active\n");
+        else {
+            printf("Activation in progress\n");
+            protocol_logic_.activate();
+
+            while(!status_.is_ready && blocking) {
+                printf("Waiting another 100ms, attempt: %d\n", iter);
+                usleep(100 * 1000);
+                read();
+                if(iter++ > 100) break;
+            }
+        }
+    };
```

Comment:
1. We should remove these `printf`s in favour of `RCLCPP_INFO`. To achieve that, please move the logs one layer higher (to the `hande_hardware_interface`).
2. To properly log the activation result, I suggest to introduce a new auxiliary type (let's say enum `ACTIVATED`/`WAITING`) which will be returned.
3. The wait logic with while and hardcoded time limit can be move to the `hande_hardware_interface`.
4. Please extract the `usleep` into utility function.

Here is my suggestion for the final activation function:
```suggestion
// [...]
inline wait() {
    usleep(100 * 1000); // 100 ms
}
// [...]
    AuxEnumOperationResult activate() {
        read();
        if (status_.is_ready)
            return AuxEnumOperationResult::ACTIVATED;
        
        protocol_logic_.activate();
        wait();
        read();
        if (status_.is_ready)
            return AuxEnumOperationResult::ACTIVATED;
        return AuxEnumOperationResult::WAITING;
    };
```

Reply:
Moving the logic totally to hardware info layer - protocol_logic_.activate() resets and sets activation flag, so it's not possible to call it more times. We also have (bool) is_ready in status so let's se how it looks in `hande_hardware_interface`.

Reply:
Is there any (for now, obvious and simple) way to pass the information for logging to the higher layer?

Reply:
We read status and get appropriate value for our information - that's actually one of few cases where we can make a use of status_ field (is_ready is what we call in hardware_interface 'activation')

Reply:
How is ready set:
`    bool is_ready() {
        return (
            gripper_status_ == GripperStatus::ACTIVATION_COMPLETE
            && activation_status_ == ActivationStatus::GRIPPER_ACTIVATION);
    };`

Reply:
Okey, seems good enough :ok_hand: